# Author Info
---
Name: **Ejaz-ur-Rehman**\
Business Unit Head | Data Analyst | Crystal Tech (Project of MUZHAB Group)\
Faculty Member-Department of Management Sciences | Al Ghazali University\
MS (Finance), PhD Scholar\
Karachi, Pakistan

![Date](https://img.shields.io/badge/Date-7--July--2026-green?logo=google-calendar)
[![Email](https://img.shields.io/badge/Email-ijazfinance%40gmail.com-blue?logo=gmail)](mailto:ijazfinance@gmail.com)
[![LinkedIn](https://img.shields.io/badge/LinkedIn-Ejaz--ur--Rehman-blue?logo=linkedin)](https://www.linkedin.com/in/ejaz-ur-rehman/)
[![GitHub](https://img.shields.io/badge/GitHub-ejazurrehman-black?logo=github)](https://github.com/ejazurrehman)

# Complete Exploratory Data Analysis with Python
## Titanic Dataset: From Raw Data to Evidence-Based Insights

**Course:** Islamic FinTech  
**Environment:** Python and Jupyter Notebook  
**Project Type:** Complete Guided Exploratory Data Analysis  
**Level:** MS IBF

---

### Purpose of This Notebook

This notebook is designed as a complete classroom journey through **Exploratory Data Analysis (EDA)**. The Titanic dataset is used because it contains a useful combination of numerical variables, categorical variables, missing values, possible outliers, group differences, and a clear outcome variable.

The objective is **not** to memorize Titanic-specific code. The objective is to learn a reusable analytical process that can later be transferred to financial, banking, FinTech, and Islamic FinTech datasets.

> **Core analytical journey:**  
> Raw Data → Understand → Audit → Question → Analyze → Visualize → Interpret → Decide → Connect → Conclude

# Notebook Roadmap

This notebook is intentionally kept as **one continuous analytical project** so that every stage connects to the next.

| Part | Main Focus | Core Question |
|---|---|---|
| **A** | Project Foundation | What are we investigating and why? |
| **B** | Python Environment and Data Loading | How do we prepare the analytical workspace? |
| **C** | Initial Data Understanding | What does the dataset contain? |
| **D** | Variable Classification | What type of data are we working with? |
| **E** | Data Quality Assessment | What problems exist in the raw data? |
| **F** | Missing Value Analysis | Where, how much, and why is data missing? |
| **G** | Duplicate and Consistency Analysis | Are records repeated or inconsistent? |
| **H** | Descriptive Statistics | What are the main numerical characteristics? |
| **I** | Univariate Analysis | What does each variable reveal individually? |
| **J** | Outlier Analysis | Which observations are unusual? |
| **K** | Bivariate Analysis | How do two variables relate? |
| **L** | Multivariate Analysis | How do several variables interact? |
| **M** | Correlation Analysis | Which numerical variables move together? |
| **N** | Feature Engineering | Can we create more useful analytical variables? |
| **O** | Integrated Analytical Questions | What deeper evidence emerges when we connect the analysis? |
| **P** | Key Findings and Conclusion | What have we learned from the evidence? |
| **Q** | Student Practice Tasks | Can students reproduce the analytical process independently? |

---

## The Repeating Learning Pattern

Every major analytical section will follow the same sequence:

1. **Concept**: What are we doing?
2. **Purpose**: Why does this step matter?
3. **Analytical Question**: What exactly are we trying to discover?
4. **Python Code**: How do we perform the analysis?
5. **Output**: What did Python return?
6. **Interpretation**: What does the result mean?
7. **Decision**: What should we do next?
8. **Connection**: How does this result lead to the next stage?

This structure is more important than any individual Python command.

# Part A: Project Foundation

## 1. What Is Exploratory Data Analysis?

>**Exploratory Data Analysis (EDA)** is the systematic process of examining, summarizing, visualizing, and questioning data before formal modelling or decision-making.

EDA helps an analyst answer questions such as:

- What does the dataset represent?
- What is the unit of analysis?
- Which variables are available?
- What types of variables are present?
- Is the data complete?
- Are there unusual or impossible values?
- What patterns appear in individual variables?
- Which variables seem related?
- Are important relationships hidden by other variables?
- What additional variables can be created?
- Which findings are strong enough to support a conclusion?

EDA is not simply the production of charts. A chart without a question and interpretation is only a graphic.

> **EDA is a reasoning process supported by statistics, Python, and visualization.**

## 2. Why EDA Comes Before Modelling

>A predictive model can process data without understanding the real-world meaning of the variables. An analyst cannot.

If we move directly to modelling, we may:

- use the wrong data type,
- ignore missing values,
- retain duplicate records,
- misunderstand categorical codes,
- allow outliers to distort results,
- create data leakage,
- misinterpret correlation,
- or produce a technically correct model based on analytically weak data.

Therefore, the analytical sequence should be:

### Step 1: Understand the problem  
### Step 2: Understand the data  
### Step 3: Assess data quality  
### Step 4: Explore individual variables  
### Step 5: Explore relationships  
### Step 6: Engineer useful features  
### Step 7: Summarize evidence  
### Step 8: Only then consider modelling

This notebook focuses on Steps 1 to 7.

## 3. The EDA Mindset

A strong analyst does not begin with:

> "Which chart should I make?"

A strong analyst begins with:

> "What do I need to understand?"

The chart, statistical measure, table, or Python method is selected **after** the analytical question is clear.

### Example

Weak approach:

```python
sns.countplot(data=df, x="Pclass")
```

Analytical approach:

**Question:** How were passengers distributed across travel classes?

**Method:** Count the number and percentage of passengers in each class.

**Visualization:** Use a count plot because `Pclass` is a categorical or ordinal variable.

**Interpretation:** Explain whether the passenger population was concentrated in a particular class.

**Connection:** Investigate whether survival outcomes differed across those classes.

The difference is not the code. The difference is the reasoning.

## 4. Project Context: The Titanic Case

>The RMS Titanic was a passenger liner that sank in 1912 after striking an iceberg during its maiden voyage. The dataset used in this project contains information about a sample of passengers and whether they survived.

>For this notebook, the historical event provides the context, but our analytical focus is the **structure of the data and the EDA process**.

We will investigate passenger characteristics such as:

- travel class,
- sex,
- age,
- family relationships,
- ticket information,
- fare,
- cabin,
- port of embarkation,
- and survival outcome.

The dataset allows us to ask a central analytical question:

> **Which passenger characteristics were associated with survival?**

>The word **associated** is important. EDA can identify patterns and relationships, but it does not automatically prove causation.

## 5. Problem Statement

The Titanic passenger dataset contains demographic, travel, family, and outcome information. However, raw data alone does not provide useful conclusions.

The analytical problem is to:

> **Systematically examine the dataset, assess its quality, understand the distribution of passenger characteristics, investigate relationships with survival, identify important interactions, and develop evidence-based conclusions.**

The project will also demonstrate how analytical decisions should emerge from evidence rather than from arbitrary cleaning or visualization choices.

## 6. Main Project Objective

>The main objective is to perform a **complete Exploratory Data Analysis of the Titanic dataset using Python** and develop a reusable EDA framework that you can later apply to financial and Islamic FinTech datasets.

### Specific Objectives

By the end of the notebook, you should be able to:

1. Explain the purpose and sequence of EDA.
2. Load and inspect a dataset in Python.
3. Identify the unit of analysis.
4. Classify variables correctly.
5. Distinguish numerical, categorical, ordinal, binary, text, and identifier variables.
6. Assess dataset dimensions and structure.
7. Detect and quantify missing values.
8. Investigate duplicate and inconsistent records.
9. Produce and interpret descriptive statistics.
10. Conduct univariate analysis.
11. Detect and interpret possible outliers.
12. Select suitable methods for bivariate analysis.
13. Conduct multivariate analysis.
14. Interpret correlation carefully.
15. Create useful engineered features.
16. Convert Python output into written analytical findings.
17. Connect one analytical finding to the next analytical question.
18. Produce a final evidence-based EDA conclusion.

## 7. Unit of Analysis

Before using Python, we must identify what one row represents.

> **Unit of Analysis: One passenger**

This means:

- each **row** represents one passenger,
- each **column** represents a recorded characteristic of that passenger,
- and the dataset combines information across many passengers.

Understanding the unit of analysis prevents major analytical errors.

### Questions to Ask in Any Dataset

Before analysis, always ask:

- What does one row represent?
- Can the same real-world entity appear more than once?
- Is the dataset at the person, transaction, account, company, country, or time-period level?
- Are all variables measured at the same level?

### Transfer to Islamic FinTech

In another project, the unit of analysis might be:

- one Islamic bank,
- one bank-year observation,
- one financing contract,
- one customer,
- one transaction,
- one Sukuk issue,
- or one Shariah-screened company.

>The correct EDA method depends partly on this unit.

## 8. Expected Variables

The commonly used Titanic training dataset contains the following variables:

| Variable | Initial Meaning | Initial Analytical Role |
|---|---|---|
| `PassengerId` | Unique passenger number | Identifier |
| `Survived` | Survival outcome | Binary target |
| `Pclass` | Passenger class | Ordinal categorical |
| `Name` | Passenger name | Text / possible feature source |
| `Sex` | Passenger sex | Categorical |
| `Age` | Passenger age | Numerical |
| `SibSp` | Siblings/spouses aboard | Discrete numerical |
| `Parch` | Parents/children aboard | Discrete numerical |
| `Ticket` | Ticket number | Identifier-like / text |
| `Fare` | Passenger fare | Continuous numerical |
| `Cabin` | Cabin number | Categorical / text |
| `Embarked` | Port of embarkation | Nominal categorical |

### Important Analytical Warning

This table is an **initial classification**, not the final answer.

For example:

- `Pclass` is stored as a number, but its values represent ordered categories.
- `PassengerId` is numeric, but calculating its mean is analytically meaningless.
- `Name` looks like free text, but titles such as Mr, Mrs, and Miss may contain useful information.
- `Cabin` has many missing values, but it may still contain useful deck information.

> **Data type in Python and analytical variable type are not always the same thing.**

## 9. The Central Analytical Question

Our main question is:

> **Which passenger characteristics were associated with survival?**

This broad question will be broken into smaller analytical questions.

### Initial Questions

- What proportion of passengers survived?
- How were passengers distributed across travel classes?
- What was the age profile of passengers?
- Was survival associated with sex?
- Was survival associated with passenger class?
- Did age differ between survivors and non-survivors?
- Did fare differ across survival groups?
- Did family structure matter?
- Did the relationship between sex and survival change across passenger classes?
- Can new features reveal patterns that are not visible in the original variables?

These questions are provisional. Good EDA generates new questions as evidence emerges.

## 10. Analytical Hypotheses Before Looking at the Results

Before examining results, it is useful to record expectations. This reduces the tendency to invent explanations only after seeing the output.

We may initially expect that:

- survival rates differed by sex,
- survival rates differed by passenger class,
- fare and passenger class were related,
- age may have influenced survival,
- passengers travelling with family may have experienced different outcomes from passengers travelling alone.

>These are **analytical expectations**, not conclusions.

>We will accept, reject, refine, or qualify them only after examining the evidence.

## 11. Complete EDA Workflow

The project will move through the following connected stages:

### Stage 1: Foundation
>Define the problem, objectives, unit of analysis, and questions.

### Stage 2: Data Acquisition
>Load the dataset and preserve the original data.

### Stage 3: Initial Inspection
>Examine rows, columns, dimensions, data types, and basic structure.

### Stage 4: Variable Classification
>Classify variables by analytical role, not merely by Python data type.

### Stage 5: Data Quality Audit
>Investigate missing values, duplicates, inconsistencies, and suspicious values.

### Stage 6: Descriptive Analysis
>Summarize central tendency, spread, frequency, and distribution.

### Stage 7: Univariate Analysis
>Understand each important variable independently.

### Stage 8: Outlier Analysis
>Identify unusual values and decide whether they are errors or valid observations.

### Stage 9: Bivariate Analysis
>Examine relationships between pairs of variables.

### Stage 10: Multivariate Analysis
>Study how relationships change when additional variables are considered.

### Stage 11: Correlation Analysis
>Measure numerical association while avoiding causal interpretation.

### Stage 12: Feature Engineering
>Create analytically useful variables from existing information.

### Stage 13: Integrated Analysis
>Connect evidence across the entire notebook.

### Stage 14: Conclusion
>Summarize findings, limitations, and next analytical steps.

## 12. Decision Rules for This Notebook

Throughout the project, we will follow several rules.

### Rule 1: Do not clean before understanding
We will not immediately delete or impute missing values.

### Rule 2: Do not visualize without a question
Every major chart must answer an analytical question.

### Rule 3: Do not interpret data types mechanically
A numeric column is not automatically a numerical analytical variable.

### Rule 4: Do not treat every unusual value as an error
An outlier may be valid and informative.

### Rule 5: Do not confuse association with causation
EDA identifies patterns, not automatic causal effects.

### Rule 6: Do not stop at Python output
Every major result requires interpretation.

### Rule 7: Connect each result to the next question
EDA should form a continuous analytical story.

## 13. Raw Data Preservation Strategy

When the dataset is loaded, we will preserve an untouched copy.

The planned structure is:

```python
df_original = df.copy()
df_eda = df.copy()
```

### Why?

- `df_original` preserves the original state.
- `df_eda` becomes the working analytical copy.
- We can compare raw and modified data.
- Accidental transformations can be checked.
- You can understand the difference between source data and prepared data.

>This practice becomes even more important in banking and FinTech projects where auditability and reproducibility matter.

## 14. What Success Looks Like

At the end of this notebook, success does **not** mean that you can reproduce Titanic charts from memory.

Success means that, when given a new dataset, you can independently ask:

1. What does one row represent?
2. What is the analytical problem?
3. Which variables are available?
4. What types of variables are they?
5. What data-quality problems exist?
6. Which summary methods are appropriate?
7. Which relationships should be investigated?
8. Which visualization matches the question?
9. What does the output mean?
10. What should be investigated next?
11. What conclusions are supported by the evidence?
12. What limitations remain?

>That is the transferable EDA skill this project is designed to develop.

# Part A Checkpoint

Before moving to Python, you should be able to answer:

- What is EDA?
- Why does EDA come before modelling?
- What is the unit of analysis in this dataset?
- What is the central analytical question?
- Why is `PassengerId` not treated like an ordinary numerical variable?
- Why should missing values not be removed immediately?
- What is the difference between a Python data type and an analytical variable type?
- Why must every visualization answer a question?
- Why does association not automatically mean causation?

> **Connection to Part B:**  
> We now understand the problem and the analytical roadmap. The next step is to prepare the Python environment and load the dataset without altering the original data.

# Part B: Python Environment and Data Loading

## 15. Purpose of This Stage

We have already defined the analytical problem. We now need to prepare a reproducible Python environment and load the data without changing its original state.

### Analytical Questions

- Which Python libraries are required?
- What role does each library play?
- How will the dataset be loaded?
- How will the raw data be preserved?
- How can we verify that loading was successful?

> **Connection from Part A:**  
> We know what we want to investigate. We can now prepare the tools and bring the data into the analytical environment.

## 16. Import the Required Libraries

We will use:

| Library | Main Purpose |
|---|---|
| `pandas` | Data loading, manipulation, grouping, and tabular analysis |
| `numpy` | Numerical operations |
| `matplotlib` | Core plotting |
| `seaborn` | Statistical visualization |


>The notebook deliberately uses a limited library set so that you focus on analytical reasoning rather than package complexity.

In [1]:
# Core data analysis libraries
import pandas as pd
import numpy as np

# Visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns

# Display settings
pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", lambda x: f"{x:,.2f}")

print("Libraries imported successfully.")

Libraries imported successfully.


### Interpretation

Successful execution means the analytical environment is ready.

### Decision

Before loading data, confirm that the Titanic CSV file is available in the same folder as this notebook.

> Recommended filename: `titanic.csv`

If your file has another name, change only the filename in the next code cell.

## 17. Load the Dataset

### Concept

`pandas.read_csv()` reads a CSV file and converts it into a DataFrame.

### Analytical Question

Did Python load the dataset correctly?

In [2]:
# Load the dataset
df = pd.read_csv("titanic.csv")

print("Dataset loaded successfully.")

Dataset loaded successfully.


## 18. Preserve the Raw Data

We should not conduct transformations directly on the only copy of the dataset.

### Strategy

- `df_original`: untouched reference copy
- `df_eda`: working copy used throughout the project

In [3]:
# Preserve the source data
df_original = df.copy()

# Create the working EDA copy
df_eda = df.copy()

print("Original and working copies created.")

Original and working copies created.


## 19. Verify the Loading Process

We will perform a basic verification before detailed analysis.

In [4]:
print("Original dataset shape:", df_original.shape)
print("Working dataset shape:", df_eda.shape)
print("Same columns:", df_original.columns.equals(df_eda.columns))

Original dataset shape: (891, 12)
Working dataset shape: (891, 12)
Same columns: True


### Interpretation and Decision

At this point, both copies should have the same dimensions and columns.

We will use `df_eda` from now onward. The original copy remains unchanged for comparison and audit purposes.

> **Connection to Part C:**  
> The data is now available. The next task is to understand its structure before making any cleaning decision.

# Part C: Initial Data Understanding

## 20. First Inspection of the Dataset

>The purpose of initial inspection is not to find every problem. It is to build the first mental model of the data.

We will examine:

- first records,
- last records,
- random records,
- dimensions,
- column names,
- data types,
- memory and non-null counts.

### Important Rule

Do not make cleaning decisions yet. First observe.

## 21. Preview the First Records

In [5]:
df_eda.head()


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.00,1,0,A/5 21171,7.25,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.00,1,0,PC 17599,71.28,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.00,0,0,STON/O2. 3101282,7.92,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.00,1,0,113803,53.10,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.00,0,0,373450,8.05,NaN,S


### Questions for Students

1 - Does one row appear to represent one passenger?\
2 - Which variables are immediately understandable?\
3 - Which variables require further investigation?\
4 - Which columns appear numerical?\
5 - Which columns appear categorical or textual?\
6 - Can missing values already be seen?

## 22. Preview the Last Records

In [7]:
df_eda.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.00,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.00,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.00,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.00,0,0,370376,7.75,NaN,Q


## 23. Inspect Random Records

Random inspection reduces the risk of forming conclusions from only the first few rows.

In [6]:
df_eda.sample(5, random_state=42)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
709,710,1,3,"Moubarek, Master. Halim Gonios (""William George"")",male,NaN,1,1,2661,15.25,NaN,C
439,440,0,2,"Kvillner, Mr. Johan Henrik Johannesson",male,31.00,0,0,C.A. 18723,10.50,NaN,S
840,841,0,3,"Alhomaki, Mr. Ilmari Rudolf",male,20.00,0,0,SOTON/O2 3101287,7.92,NaN,S
720,721,1,2,"Harper, Miss. Annie Jessie ""Nina""",female,6.00,0,1,248727,33.00,NaN,S
39,40,1,3,"Nicola-Yarred, Miss. Jamila",female,14.00,1,0,2651,11.24,NaN,C


## 24. Dataset Dimensions

In [8]:
rows, columns = df_eda.shape

print(f"Number of rows: {rows}")
print(f"Number of columns: {columns}")

Number of rows: 891
Number of columns: 12


### Interpretation

- Rows represent observations.
- Columns represent variables.

For the standard Titanic training dataset, the expected unit of analysis is one passenger.

## 25. Column Names

In [9]:
df_eda.columns.tolist()

['PassengerId',
 'Survived',
 'Pclass',
 'Name',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Ticket',
 'Fare',
 'Cabin',
 'Embarked']

## 26. Data Types

In [10]:
df_eda.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

### Critical Interpretation

Python data types describe **storage**, not necessarily analytical meaning.

Examples:

- `PassengerId` may be stored as an integer but functions as an identifier.
- `Pclass` may be stored as an integer but represents ordered categories.
- `Survived` may be stored as an integer but represents a binary outcome.

We will formally classify variables in Part D.

### Important Definition:

1. int64
   
Definition

>int64 means 64-bit integer.An integer is a whole number without a decimal point.

Simple Examples
- 10
- 25
- 100
- -5
- 0

These are integers because they do not contain decimal values.

Titanic Examples
- PassengerId
- Survived
- Pclass
- SibSp
- Parch

For example:

PassengerId = 1\
Survived = 0\
Pclass = 3\
SibSp = 1\
Parch = 0

Pandas may store these columns as:

int64\
What Does 64 Mean?

>The 64 refers to the amount of computer memory used to store the value.

For beginners, the main point is:

>int64 stores whole numbers.

Important Analytical Warning

>A column stored as int64 is not automatically a numerical analytical variable.

For example:

| Variable      | Python Data Type | Analytical Meaning        |
| ------------- | ---------------- | ------------------------- |
| `PassengerId` | `int64`          | Identifier                |
| `Survived`    | `int64`          | Binary categorical target |
| `Pclass`      | `int64`          | Ordinal categorical       |
| `SibSp`       | `int64`          | Discrete numerical        |

This is a very important distinction.

2. float64

Definition

>float64 means 64-bit floating-point number.A float is a number that can contain a decimal point.

Simple Examples
- 10.5
- 25.75
- 3.14
- 100.00

Titanic Examples
- Age
- Fare

Example values:

Age = 22.0\
Age = 38.0\
Fare = 7.25\
Fare = 71.2833

Pandas may store these columns as:

float64\
Why Is Age a Float?

>“Age is usually a whole number. Why is it float64?”

There are two reasons.

>First, the Titanic dataset contains some fractional ages, especially for infants.

For example:

Age = 0.75

>Second, Pandas traditionally uses floating-point storage when a numerical column contains missing values such as:

NaN

Therefore:

Age

is stored as:

float64

Key Point

>float64 stores numerical values that may contain decimal points.

3. str or Object

Definition

>str means string.A string is a sequence of text characters.

It can contain:

- letters,
- words,
- spaces,
- numbers,
- symbols,
or combinations of them.

Simple Examples
- "Ejaz"
- "Pakistan"
- "Male"
- "Financew-101"

All of these are strings.

Titanic Examples

The following variables contain text:

- Name
- Sex
- Ticket
- Cabin
- Embarked

Example values:

Name = Braund, Mr. Owen Harris\
Sex = male\
Ticket = A/5 21171\
Cabin = C85\
Embarked = S\

Important Point

>Even if a string contains numbers, Python may still treat it as text.

For example:

"PC 17599"

is text, not a numerical value.

We cannot meaningfully calculate:

- Average Ticket Number

**Important Pandas Note: str vs object**

You may notice that df.dtypes shows:

object

instead of:

str

For example:

Name        object\
Sex         object\
Ticket      object\
Cabin       object\
Embarked    object

In many Pandas datasets, text columns are stored as:

object

So, for beginner-level understanding:

>object usually means the column contains text or strings.However, technically, an object column can contain mixed Python objects. Therefore, we should inspect the values before making a final analytical decision.

Quick Comparison

| Data Type | Meaning                             | Example  | Titanic Variable        |
| --------- | ----------------------------------- | -------- | ----------------------- |
| `int64`   | Whole numbers                       | `25`     | `PassengerId`, `Pclass` |
| `float64` | Numbers with possible decimals      | `22.5`   | `Age`, `Fare`           |
| `str`     | Text                                | `"male"` | `Name`, `Sex`           |
| `object`  | Common Pandas storage type for text | `"C85"`  | `Ticket`, `Cabin`       |

Final Learning Point

>Python data type tells us how data is stored. Analytical variable type tells us how data should be analyzed.

For example:
| Variable      | Python Data Type | Analytical Variable Type  |
| ------------- | ---------------- | ------------------------- |
| `PassengerId` | `int64`          | Identifier                |
| `Survived`    | `int64`          | Binary categorical target |
| `Pclass`      | `int64`          | Ordinal categorical       |
| `Age`         | `float64`        | Continuous numerical      |
| `Sex`         | `str` / `object` | Nominal categorical       |
| `Ticket`      | `str` / `object` | Identifier-like text      |


## 27. Complete Structural Summary

In [11]:
df_eda.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


### What `info()` Helps Us See

- total observations,
- column names,
- non-null counts,
- data types,
- approximate memory use.

### Initial Questions

1 - Which columns have fewer non-null values than the total number of rows?

2 - Which variables may contain missing data?

3 - Are any data types surprising?

1. Which columns have fewer non-null values than the total number of rows?

The dataset contains 891 rows. Three columns have fewer than 891 non-null values:

- Age: 714 non-null values
- Cabin: 204 non-null values
- Embarked: 889 non-null values

2. Which variables may contain missing data?

The variables that may contain missing data are:

- Age
- Cabin
- Embarked

>The exact number and percentage of missing values will be calculated later.

3. Are any data types surprising?

Yes. Some variables are stored as numbers but should not be analyzed as ordinary numerical variables:

- PassengerId: int64, but analytically an identifier
- Survived: int64, but analytically a binary categorical target
- Pclass: int64, but analytically an ordinal categorical variable

>Key observation: Python data type tells us how data is stored, while analytical variable type tells us how the data should be analyzed.



## 28. Number of Unique Values

Cardinality means the number of distinct values in a variable.

In [12]:
df_eda.nunique().sort_values()

Survived         2
Sex              2
Pclass           3
Embarked         3
Parch            7
SibSp            7
Age             88
Cabin          147
Fare           248
Ticket         681
PassengerId    891
Name           891
dtype: int64

### Why Cardinality Matters

Cardinality helps distinguish:

- binary variables,
- low-cardinality categories,
- high-cardinality text variables,
- possible identifiers.

However, cardinality alone is not enough. Context remains necessary.

## 29. Initial Inspection Summary Table

In [13]:
initial_summary = pd.DataFrame({
    "Data Type": df_eda.dtypes.astype(str),
    "Non-Null Count": df_eda.notna().sum(),
    "Missing Count": df_eda.isna().sum(),
    "Unique Values": df_eda.nunique()
})

initial_summary

,Data Type,Non-Null Count,Missing Count,Unique Values
PassengerId,int64,891,0,891
Survived,int64,891,0,2
Pclass,int64,891,0,3
Name,str,891,0,891
Sex,str,891,0,2
Age,float64,714,177,88
SibSp,int64,891,0,7
Parch,int64,891,0,7
Ticket,str,891,0,681
Fare,float64,891,0,248


### Decision

We now have enough structural information to classify variables properly.

> **Connection to Part D:**  
> Before calculating statistics or drawing charts, we must decide what each variable analytically represents.

### Initial Data Understanding

>The dataset contains **891 passengers and 19 original variables**. One row represents one passenger, confirming the unit of analysis defined in Part A.

>The first inspection shows a mixture of numerical, categorical, text, and identifier-like variables. `Age`, `Cabin`, and `Embarked` contain missing values. The structural inspection also confirms that Python storage types do not fully describe analytical meaning. For example, `Pclass` and `Survived` are stored numerically but should not be treated as ordinary continuous numerical variables.

>**Decision:** The dataset loaded correctly and is suitable for EDA. No cleaning action should be taken until variables are classified and data-quality issues are quantified.

# Part D: Variable Classification

## 30. Why Variable Classification Matters

The analytical method depends on variable type.

A mean is meaningful for `Age`, but not for `PassengerId`. A histogram is appropriate for `Fare`, but usually not for `Embarked`.

### Main Analytical Categories

- Identifier
- Binary categorical
- Nominal categorical
- Ordinal categorical
- Discrete numerical
- Continuous numerical
- Text
- Target variable

## 31. Create the Analytical Variable Classification

### Key Learning Point

The correct question is not:

> "Is this column numeric?"

The better question is:

> "What does this variable represent, and which analytical operations are meaningful?"

### Decision

We now know how variables should be treated. The next task is to audit data quality systematically.

### Variable Classification

>The main analytical distinction is between **storage type** and **meaning**. `PassengerId` is an integer but serves only as an identifier. `Pclass` is an ordered category, not a continuous measurement. `Survived` is a binary target. `Age` and `Fare` are continuous numerical variables, while `SibSp` and `Parch` are discrete counts.

>**Decision:** Statistical methods and visualizations will be selected according to analytical type, not simply according to the output of `df.dtypes`.